pallets/order-book/src/lib.rs

```diff
@@ -592,6 +660,198 @@ impl<T: Config> Pallet<T> {
+                error: Error::<T>::UnknownOrderBook.into(),
+            });
+            return;
+        };
+
+        if let Err(error) = order_book.cancel_limit_order_unchecked::<Self>(order, data_layer) {
+            debug_assert!(
+                false,
+                "expiration resulted in error, this must not happen: {:?}",
+                error
+            );
+            Self::deposit_event(Event::<T>::ExpirationFailure {
+                order_book_id: order_book_id.clone(),
+                order_id: order_id.clone(),
+                error,
+            });
+        }
+    }
+
+    /// Expire orders that are scheduled to expire at `block`.
+    /// `weight` is used to track weight spent on the expirations, so that
+    /// it doesn't accidentally spend weight of the entire block (or even more).
+    ///
+    /// Returns `true` if all expirations were processed and `false` if some expirations
+    /// need to be retried when more weight is available.
+    fn service_block(
+        data_layer: &mut CacheDataLayer<T>,
+        block: T::BlockNumber,
+        weight: &mut WeightMeter,
+    ) -> bool {
+        if !weight.check_accrue(<T as Config>::WeightInfo::service_block_base()) {
+            return false;
+        }
+
+        let mut expirations = <ExpirationsAgenda<T>>::take(block);
+        if expirations.is_empty() {
+            return true;
+        }
+        // how many we can service with remaining weight;
+        // the weight is consumed right away
+        let to_service = check_accrue_n(
+            weight,
+            <T as Config>::WeightInfo::service_single_expiration(),
+            expirations.len() as u64,
+        );
+        let postponed = expirations.len() as u64 - to_service;
+        let mut serviced = 0;
+        while let Some((order_book_id, order_id)) = expirations.last() {
+            if serviced >= to_service {
+                break;
+            }
+            Self::service_single_expiration(data_layer, order_book_id, order_id);
+            serviced += 1;
+            expirations.pop();
+        }
+        if postponed != 0 {
+            // Will later continue from this block
+            <ExpirationsAgenda<T>>::insert(block, expirations);
+        }
+        postponed == 0
+    }
+}
+
+#[derive(Debug)]
+pub enum ScheduleError {
+    /// Expiration schedule for this block is full
+    BlockScheduleFull,
+}
+
+impl<T: Config> Into<Error<T>> for ScheduleError {
+    fn into(self) -> Error<T> {
+        match self {
+            ScheduleError::BlockScheduleFull => Error::<T>::BlockScheduleFull,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum UnscheduleError {
+    /// Could not find expiration in given block schedule
+    ExpirationNotFound,
+}
+
+impl<T: Config> Into<Error<T>> for UnscheduleError {
+    fn into(self) -> Error<T> {
+        match self {
+            UnscheduleError::ExpirationNotFound => Error::<T>::ExpirationNotFound,
+        }
+    }
+}
```

Comment:
(discussed in dms, replying for the record)
I thought that it would be nice to know that scheduling produces only a single error variant (especially it seemed useful in `on_initialize` part). But it turned out to be unused, so removing the types.